Bogus extra error about unsized variable (the trait bound `str: std::marker::Sized` is not satisfied)

Given code like this (not complete):

```rust
extern crate serde_json;
let g1 = vec![0; 16];
let json1 = serde_json::to_pretty_string(&g1).unwrap();
let foo: Foo = serde_json::from_str(&json1).unwrap();
```

The first error is correct but the second is nonsense:

```rust
error[E0425]: cannot find function `to_pretty_string` in module `serde_json`
   --> serialization-tests/tests/serialization.rs:498:35
    |
498 |         let json1 = serde_json::to_pretty_string(&g1).unwrap();
    |                                 ^^^^^^^^^^^^^^^^ not found in `serde_json`

error[E0277]: the trait bound `str: std::marker::Sized` is not satisfied
   --> serialization-tests/tests/serialization.rs:498:13
    |
498 |         let json1 = serde_json::to_pretty_string(&g1).unwrap();
    |             ^^^^^ `str` does not have a constant size known at compile-time
    |
    = help: the trait `std::marker::Sized` is not implemented for `str`
    = note: all local variables must have a statically known size
```


There's an extra error here. `json1` is used in a way (with the `from_str` function) that it should be a `str` or something that derefs to `str`. The second error message supposes it is a `str`, which is nonsense that the user doesn't intend at all. (If the user would only find the right name of the function, they would get a String in the json variable and the next line compiles fine.)

--- a/src/librustc_typeck/check/coercion.rs
+++ b/src/librustc_typeck/check/coercion.rs
@@ -163,7 +163,7 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
 
         // Just ignore error types.
         if a.references_error() || b.references_error() {
-            return success(vec![], b, vec![]);
+            return success(vec![], self.fcx.tcx.types.err, vec![]);
         }
 
         if a.is_never() {
@@ -821,7 +821,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
         let (adjustments, _) = self.register_infer_ok_obligations(ok);
         self.apply_adjustments(expr, adjustments);
-        Ok(target)
+        if expr_ty.references_error() {
+            Ok(self.tcx.types.err)
+        } else {
+            Ok(target)
+        }
     }
 
     /// Same as `try_coerce()`, but without side-effects.

--- a/src/librustc_typeck/check/coercion.rs
+++ b/src/librustc_typeck/check/coercion.rs
@@ -821,11 +821,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
         let (adjustments, _) = self.register_infer_ok_obligations(ok);
         self.apply_adjustments(expr, adjustments);
-        if expr_ty.references_error() {
-            Ok(self.tcx.types.err)
+        Ok(if expr_ty.references_error() {
+            self.tcx.types.err
         } else {
-            Ok(target)
-        }
+            target
+        })
     }
 
     /// Same as `try_coerce()`, but without side-effects.

--- a/src/librustc_typeck/check/mod.rs
+++ b/src/librustc_typeck/check/mod.rs
@@ -153,7 +153,7 @@ use self::method::{MethodCallee, SelfSource};
 use self::TupleArgumentsFlag::*;
 
 /// The type of a local binding, including the revealed type for anon types.
-#[derive(Copy, Clone)]
+#[derive(Copy, Clone, Debug)]
 pub struct LocalTy<'tcx> {
     decl_ty: Ty<'tcx>,
     revealed_ty: Ty<'tcx>
@@ -3752,14 +3752,35 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         if let Some(ref init) = local.init {
             let init_ty = self.check_decl_initializer(local, &init);
             if init_ty.references_error() {
+                // Override the types everywhere with `types.err` to avoid knock down errors.
                 self.write_ty(local.hir_id, init_ty);
+                self.write_ty(local.pat.hir_id, init_ty);
+                self.locals.borrow_mut().insert(local.hir_id, LocalTy {
+                    decl_ty: t,
+                    revealed_ty: init_ty,
+                });
+                self.locals.borrow_mut().insert(local.pat.hir_id, LocalTy {
+                    decl_ty: t,
+                    revealed_ty: init_ty,
+                });
             }
         }
 
         self.check_pat_top(&local.pat, t, None);
         let pat_ty = self.node_ty(local.pat.hir_id);
+        debug!("check_decl_local pat_ty {:?}", pat_ty);
         if pat_ty.references_error() {
+            // Override the types everywhere with `types.err` to avoid knock down errors.
             self.write_ty(local.hir_id, pat_ty);
+            self.write_ty(local.pat.hir_id, pat_ty);
+            self.locals.borrow_mut().insert(local.hir_id, LocalTy {
+                decl_ty: t,
+                revealed_ty: pat_ty,
+            });
+            self.locals.borrow_mut().insert(local.pat.hir_id, LocalTy {
+                decl_ty: t,
+                revealed_ty: pat_ty,
+            });
         }
     }
 

--- a/src/librustc_typeck/check/mod.rs
+++ b/src/librustc_typeck/check/mod.rs
@@ -3751,36 +3751,25 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
         if let Some(ref init) = local.init {
             let init_ty = self.check_decl_initializer(local, &init);
-            if init_ty.references_error() {
-                // Override the types everywhere with `types.err` to avoid knock down errors.
-                self.write_ty(local.hir_id, init_ty);
-                self.write_ty(local.pat.hir_id, init_ty);
-                self.locals.borrow_mut().insert(local.hir_id, LocalTy {
-                    decl_ty: t,
-                    revealed_ty: init_ty,
-                });
-                self.locals.borrow_mut().insert(local.pat.hir_id, LocalTy {
-                    decl_ty: t,
-                    revealed_ty: init_ty,
-                });
-            }
+            self.overwrite_local_ty_if_err(local, t, init_ty);
         }
 
         self.check_pat_top(&local.pat, t, None);
         let pat_ty = self.node_ty(local.pat.hir_id);
-        debug!("check_decl_local pat_ty {:?}", pat_ty);
-        if pat_ty.references_error() {
+        self.overwrite_local_ty_if_err(local, t, pat_ty);
+    }
+
+    fn overwrite_local_ty_if_err(&self, local: &'tcx hir::Local, decl_ty: Ty<'tcx>, ty: Ty<'tcx>) {
+        if ty.references_error() {
             // Override the types everywhere with `types.err` to avoid knock down errors.
-            self.write_ty(local.hir_id, pat_ty);
-            self.write_ty(local.pat.hir_id, pat_ty);
-            self.locals.borrow_mut().insert(local.hir_id, LocalTy {
-                decl_ty: t,
-                revealed_ty: pat_ty,
-            });
-            self.locals.borrow_mut().insert(local.pat.hir_id, LocalTy {
-                decl_ty: t,
-                revealed_ty: pat_ty,
-            });
+            self.write_ty(local.hir_id, ty);
+            self.write_ty(local.pat.hir_id, ty);
+            let local_ty = LocalTy {
+                decl_ty,
+                revealed_ty: ty,
+            };
+            self.locals.borrow_mut().insert(local.hir_id, local_ty);
+            self.locals.borrow_mut().insert(local.pat.hir_id, local_ty);
         }
     }
 

--- a/src/test/ui/issues/issue-33575.rs
+++ b/src/test/ui/issues/issue-33575.rs
@@ -1,0 +1,4 @@
+fn main() {
+    let baz = ().foo(); //~ ERROR no method named `foo` found for type `()` in the current scope
+    <i32 as std::str::FromStr>::from_str(&baz); // No complains about `str` being unsized
+}

--- a/src/test/ui/issues/issue-33575.rs
+++ b/src/test/ui/issues/issue-33575.rs
@@ -1,4 +1,4 @@
 fn main() {
     let baz = ().foo(); //~ ERROR no method named `foo` found for type `()` in the current scope
-    <i32 as std::str::FromStr>::from_str(&baz); // No complains about `str` being unsized
+    <i32 as std::str::FromStr>::from_str(&baz); // No complaints about `str` being unsized
 }

--- a/src/test/ui/issues/issue-33575.stderr
+++ b/src/test/ui/issues/issue-33575.stderr
@@ -1,0 +1,9 @@
+error[E0599]: no method named `foo` found for type `()` in the current scope
+  --> $DIR/issue-33575.rs:2:18
+   |
+LL |     let baz = ().foo();
+   |                  ^^^ method not found in `()`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0599`.

